src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs

```diff
@@ -637,57 +639,40 @@ private unsafe string[] GetValueNamesCore(int values)
                                 throw new IOException(SR.Arg_RegGetOverflowBug, e);
                             }
                         }
-                        char[] blob = new char[datasize / 2];
 
-                        Interop.Advapi32.RegQueryValueEx(_hkey, name, null, ref type, blob, ref datasize);
-                        if (blob.Length > 0 && blob[blob.Length - 1] == (char)0)
-                        {
-                            data = new string(blob, 0, blob.Length - 1);
-                        }
-                        else
+                        char[]? pooledArray = null;
+                        int blobLength = datasize / 2;
```

Comment:
Now that this is using stackallocation and pooled buffer, it would be nice to avoid the two RegQueryValueEx calls for common cases. We can stackallocate the buffer and pass it into the first RegQueryValueEx call. If the buffer was big enough, we are done. If the buffer was not big enough, we retry with properly sized buffer.

Reply:
In making this change, we appear to not properly handle the cases where the key is changed concurrently with the operation, e.g. there's no retry for REG_BINARY.  Is there a reason that's not needed?

Reply:
Ok, I'm just going to rewrite this method.

Reply:
@jkotas, we have a handful of tests that are trying to validate roundtripping of corrupted data, e.g. storing new byte[] { 6, 5, 6 } into a REG_SZ key and expecting to get back a string containing { 0x0506, 6 }.  But RegQueryValueEx is showing some interesting behaviors here.  If you pass in a buffer length of 3 or 4 to retrieve that data, you get it back, but any buffer length larger than 4, and the 3rd byte is zero'd such that we instead get back just { 0x0506 }.  As such I don't see a good way to retain the existing behavior in these (presumably all corrupt) cases while also supporting making a single call to RegQueryValueEx.  I'm tempted to say it's an acceptable change: corrupt in, corrupt out.  Opinions?